src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,300 @@
   }
-#endif
 
-  // Try to load the file; but if it's raw_binary, it could be a problem.
-  if (loadType == FileType::RawBinary)
-    Log::Warn << "Loading '" << filename << "' as " << stringType << "; "
-        << "but this may not be the actual filetype!" << std::endl;
-  else
-    Log::Info << "Loading '" << filename << "' as " << stringType << ".  "
-        << std::flush;
+  return success;
+}
 
-  // We can't use the stream if the type is HDF5.
-  bool success;
-  LoadCSV loader;
+template<typename MatType, typename DataOptionsType>
+bool Load(const std::string& filename,
+          MatType& matrix,
+          const DataOptionsType& opts,
+          std::enable_if_t<IsArma<MatType>::value || IsSparseMat<MatType>::value>*,
+          std::enable_if_t<!std::is_same_v<DataOptionsType, bool>>*)
+{
+  DataOptionsType tmpOpts(opts);
+  return Load(filename, matrix, tmpOpts);
+}
 
-  if (loadType != FileType::HDF5Binary)
+template<typename MatType>
+bool LoadMatrix(const std::string& filename,
+                MatType& matrix,
+                std::fstream& stream,
+                TextOptions& txtOpts)
+{
+  bool success = false;
+  if constexpr (IsSparseMat<MatType>::value)
   {
-    if (loadType == FileType::CSVASCII)
-      success = loader.LoadNumericCSV(matrix, stream);
-    else
-      success = matrix.load(stream, ToArmaFileType(loadType));
+    success = LoadSparse(filename, matrix, txtOpts, stream);
   }
-  else
-    success = matrix.load(filename, ToArmaFileType(loadType));
-
-  if (!success)
+  else if (txtOpts.Categorical() ||
+      (txtOpts.Format() == FileType::ArffASCII))
   {
-    Log::Info << std::endl;
-    Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Loading from '" << filename << "' failed." << std::endl;
-    else
-      Log::Warn << "Loading from '" << filename << "' failed." << std::endl;
-
-    return false;
+    success = LoadCategorical(filename, matrix, txtOpts);
+  }
+  else if constexpr (IsCol<MatType>::value)
+  {
+    success = LoadCol(filename, matrix, txtOpts, stream);
+  }
+  else if constexpr (IsRow<MatType>::value)
+  {
+    success = LoadRow(filename, matrix, txtOpts, stream);
+  }
+  else if constexpr (IsDense<MatType>::value)
+  {
+    success = LoadDense(filename, matrix, txtOpts, stream);
   }
   else
-    Log::Info << "Size is " << (transpose ? matrix.n_cols : matrix.n_rows)
-        << " x " << (transpose ? matrix.n_rows : matrix.n_cols) << ".\n";
-
-  // Now transpose the matrix, if necessary.
-  if (transpose)
   {
-    success = inplace_transpose(matrix, fatal);
+    throw std::runtime_error("Matrix type is not detected!");
   }
-
-  Timer::Stop("loading_data");
-
-  // Finally, return the success indicator.
   return success;
 }
 
-// Load with mappings.  Unfortunately we have to implement this ourselves.
-template<typename eT, typename PolicyType>
+// This is the function that the user is supposed to call.
+// Other functions of this class should be labelled as private.
```

Comment:
I'm not sure what you mean here, these are all bare functions and not in a class?

Reply:
I am referring to LoadDense,  LoadSparse, LoadCategorical, etc..